### PR TITLE
Deprecate SelfResolvingDependency

### PR DESCRIPTION
**Pull request by jvandort**
SelfResolvingDependency is archaic. Dependencies should not be resolved without a Configuration.

In Gradle 9.0, ProjectDependency and FileCollectionDependency will no longer implement this interface. 

A ProjectDependency is a SelfResolvingDependency, but in practice it is not resolved that way by the dependency resolution engine. In order to support the self-resolving methods, a ProjectDependency needs an actual Project instance (or an instance to the project registry) This should not be necessary. A project dependency should be a simple reference to the identifier of a project and no more.

A FileCollectionDependency already exposes its FileCollection, so existing calls to the resolve method are easily replaced.

SelfResolvingDependencyInternal is still used by DefaultFileCollectionDependency, but cannot be renamed since KMP uses that internal interface to implement NPM dependencies.

### Reviewing cheatsheet

Before merging the PR, comments starting with 
- ❌ ❓**must** be fixed
- 🤔 💅 **should** be fixed
- 💭 **may** be fixed
- 🎉 celebrate happy things

### CODE REVIEW

**Comment by jvandort** [comment]
Blocked by https://issuetracker.google.com/issues/316294717

**Comment by jvandort** [comment]
@bot-gradle test ast

**Comment by jvandort** [comment]
@bot-gradle test ast

**Comment by jvandort** [comment]
@bot-gradle test ast

**Comment by jvandort** [comment]
@bot-gradle test this

**Review by lkasso — APPROVED** [review]
Docs LGTM

**Review by eskatos — APPROVED** [review]
Kotlin DSL changes LGTM

File: platforms/documentation/docs/src/docs/userguide/releases/upgrading/upgrading_version_8.adoc
Comment (big-guy) [inline platforms/documentation/docs/src/docs/userguide/releases/upgrading/upgrading_version_8.adoc:212]: ❌ This is in the wrong version section now

File: platforms/software/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/DependencyNotationIntegrationSpec.groovy
Comment (big-guy) [inline platforms/software/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/DependencyNotationIntegrationSpec.groovy:37]: 🤔 Is this really supposed to be a directly created self resolving dependency or can it just be `files(...)`?

File: platforms/core-configuration/configuration-cache/src/integTest/groovy/org/gradle/configurationcache/isolated/IsolatedProjectsAccessFromGroovyDslIntegrationTest.groovy
Comment (big-guy) [inline platforms/core-configuration/configuration-cache/src/integTest/groovy/org/gradle/configurationcache/isolated/IsolatedProjectsAccessFromGroovyDslIntegrationTest.groovy:709]: Why was this removed?

File: subprojects/core/src/main/java/org/gradle/api/internal/artifacts/dependencies/ProjectDependencyInternal.java
Comment (big-guy) [inline subprojects/core/src/main/java/org/gradle/api/internal/artifacts/dependencies/ProjectDependencyInternal.java:33]: 🤔 This looks a little out of place. I don't think any of the other Dependency types have a component identifier in them. 

Dependency is the input to the resolution process and ComponentIdentifiers are closer to the result of the process.  Could we get by by using the existing `getDependencyProject`?

e.g., `DefaultProjectComponentSelector.newSelector(projectDependency.getDependencyProject());`

File: subprojects/core/src/main/java/org/gradle/api/internal/artifacts/dependencies/DefaultFileCollectionDependency.java
Comment (big-guy) [inline subprojects/core/src/main/java/org/gradle/api/internal/artifacts/dependencies/DefaultFileCollectionDependency.java:132]: `Boolean.getBoolean(...)` does what you want

**Review by big-guy — CHANGES_REQUESTED** [review]

File: subprojects/core/src/main/java/org/gradle/api/internal/artifacts/dependencies/ProjectDependencyInternal.java
Comment (jvandort) [reply subprojects/core/src/main/java/org/gradle/api/internal/artifacts/dependencies/ProjectDependencyInternal.java:33]: One of the major motivations behind this deprecation was to remove usages of `getDependencyProject`. 

IMO, I don't think a project dependency should have a `getDependencyProject` method on it at all. Especially in an "Isolated Projects" world. I should not need an instance of a project to create a project dependency. 

A project dependency IMO should just be a wrapper around some unique identifier for a project. According to `ProjectStateRegistry`, projects are uniquely identified either by a `ProjectComponentIdentifier` or a `Path` representing the "identity path" (the build path + the project path within that build).

To me, this is just exposing a unique identifier for the project being referenced by the dependency. 

If I could choose here, I would prefer to use the `getIdentityPath` already exposed from the project dependency, but due to the nature of `ProjectComponentSelector`, we also need to expose the "raw" build path and project path. 

Since these changes are not necessary for the deprecation, I've reverted this, but I think we should be moving _someway_ in this direction. I think we should avoid using `getDependencyProject` and move towards deprecating the method.

File: platforms/software/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/DependencyNotationIntegrationSpec.groovy
Comment (jvandort) [reply platforms/software/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/DependencyNotationIntegrationSpec.groovy:37]: This does not need to be directly created.

File: platforms/core-configuration/configuration-cache/src/integTest/groovy/org/gradle/configurationcache/isolated/IsolatedProjectsAccessFromGroovyDslIntegrationTest.groovy
Comment (jvandort) [reply platforms/core-configuration/configuration-cache/src/integTest/groovy/org/gradle/configurationcache/isolated/IsolatedProjectsAccessFromGroovyDslIntegrationTest.groovy:709]: This causes a deprecation warning and it appears I was too lazy to add the `expectDocumentedDeprecationWarning` call

File: platforms/documentation/docs/src/docs/userguide/releases/upgrading/upgrading_version_8.adoc
Comment (jvandort) [reply platforms/documentation/docs/src/docs/userguide/releases/upgrading/upgrading_version_8.adoc]: Moved

File: subprojects/core/src/main/java/org/gradle/api/internal/artifacts/dependencies/DefaultFileCollectionDependency.java
Comment (jvandort) [reply subprojects/core/src/main/java/org/gradle/api/internal/artifacts/dependencies/DefaultFileCollectionDependency.java]: Changed

File: platforms/core-configuration/configuration-cache/src/integTest/groovy/org/gradle/configurationcache/isolated/IsolatedProjectsAccessFromGroovyDslIntegrationTest.groovy
Comment (tresat) [inline platforms/core-configuration/configuration-cache/src/integTest/groovy/org/gradle/configurationcache/isolated/IsolatedProjectsAccessFromGroovyDslIntegrationTest.groovy:693]:
```suggestion
            executer.expectDocumentedDeprecationWarning("Accessing the build dependencies of a project dependency has been deprecated. This will fail with an error in Gradle 9.0. Add the dependency to a resolvable configuration and use the configuration to track task dependencies. Consult the upgrading guide for further information: https://docs.gradle.org/current/userguide/upgrading_version_8.html#deprecate_self_resolving_dependency")
```

File: platforms/documentation/docs/src/docs/userguide/releases/upgrading/upgrading_version_8.adoc
Comment (tresat) [inline platforms/documentation/docs/src/docs/userguide/releases/upgrading/upgrading_version_8.adoc:58]: 🤔 What's the replacement for this method?  It's not demoed below.

File: platforms/documentation/docs/src/docs/userguide/releases/upgrading/upgrading_version_8.adoc
Comment (tresat) [inline platforms/documentation/docs/src/docs/userguide/releases/upgrading/upgrading_version_8.adoc:90]: 🤔 I think this would be clearer with an additional comment like "iterating through the files implicitly resolves the dependencies.

I also really think this should be split into 2 tasks "resolveDeprecated" and "resolve" (showing the replacement) that both do something with the resolution result..  This doesn't show the equivalent operation (iterating the file names) using the old way and new way.

File: platforms/software/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/SelfResolvingDependencyIntegrationTest.groovy
Comment (tresat) [inline platforms/software/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/SelfResolvingDependencyIntegrationTest.groovy:52]: 🤔 This would be clearer splitting the deprecation message generating `resolve` into a separate test than the replacement.

Or maybe as a parameterized test, making clear to only expect the dep warnings using the older methods.

File: platforms/software/dependency-management/src/main/java/org/gradle/api/internal/artifacts/configurations/DefaultConfiguration.java
Comment (tresat) [inline platforms/software/dependency-management/src/main/java/org/gradle/api/internal/artifacts/configurations/DefaultConfiguration.java:915]: ❓ Why change this constructor from accepting a `DependencySet` and doing the filtering to only `ProjectDependency`s internally?

File: platforms/software/dependency-management/src/main/java/org/gradle/api/internal/artifacts/configurations/TasksFromProjectDependencies.java
Comment (tresat) [inline platforms/software/dependency-management/src/main/java/org/gradle/api/internal/artifacts/configurations/TasksFromProjectDependencies.java:54]: ❗ Non `private static` methods called from constructor of a non-`final` class are trouble.

The only field used here is `taskName`.  Which is only read in a test.  Can this field be deleted completely and just passed here as an argument?

File: platforms/software/dependency-management/src/main/java/org/gradle/api/internal/artifacts/dependencies/DefaultProjectDependency.java
Comment (tresat) [inline platforms/software/dependency-management/src/main/java/org/gradle/api/internal/artifacts/dependencies/DefaultProjectDependency.java:115]: 🤔 I think it would be better to include the name of the project dep that is being directly resolved here.  Yes, this might produce multiple similar deprecation warnings, but it will greatly benefit the maintainer to immediately see which project dep is responsible - this is first question anyone will have when they see this warning.

Comment (tresat) [reply platforms/software/dependency-management/src/main/java/org/gradle/api/internal/artifacts/dependencies/DefaultProjectDependency.java:115]: Also for `getBuildDependencies()`.

File: platforms/software/dependency-management/src/main/java/org/gradle/api/internal/catalog/DelegatingProjectDependency.java
Comment (tresat) [inline platforms/software/dependency-management/src/main/java/org/gradle/api/internal/catalog/DelegatingProjectDependency.java:194]: 💅 Suppressing the deprecation warning isn't necessary if these methods themselves are deprecated.  Applies throughout

File: subprojects/core-api/src/main/java/org/gradle/api/artifacts/FileCollectionDependency.java
Comment (tresat) [inline subprojects/core-api/src/main/java/org/gradle/api/artifacts/FileCollectionDependency.java:51]: 💅 Shouldn't these be marked `@Deprecated` as well, to be consistent with `getBuildDependencies()`?  Applies throughout.

File: subprojects/core/src/main/java/org/gradle/api/internal/artifacts/CachingDependencyResolveContext.java
Comment (tresat) [inline subprojects/core/src/main/java/org/gradle/api/internal/artifacts/CachingDependencyResolveContext.java:63]: 🤔 This class is internal, can you just split this into 2 overloaded methods that are strongly-typed?

File: subprojects/core/src/main/java/org/gradle/api/internal/artifacts/dependencies/SelfResolvingDependencyInternal.java
Comment (tresat) [inline subprojects/core/src/main/java/org/gradle/api/internal/artifacts/dependencies/SelfResolvingDependencyInternal.java:23]: 💅 This should be javadoc, to be more visible to readers.

File: subprojects/core/src/main/java/org/gradle/api/internal/artifacts/dependencies/DefaultFileCollectionDependency.java
Comment (tresat) [inline subprojects/core/src/main/java/org/gradle/api/internal/artifacts/dependencies/DefaultFileCollectionDependency.java:104]: ❓ Does this emit always because it's not used by AGP?

File: subprojects/core/src/main/java/org/gradle/api/internal/artifacts/dependencies/DefaultFileCollectionDependency.java
Comment (tresat) [inline subprojects/core/src/main/java/org/gradle/api/internal/artifacts/dependencies/DefaultFileCollectionDependency.java:91]: 🤔 Is there any way to make it clearer which dependency this is talking about?

File: platforms/software/dependency-management/src/main/java/org/gradle/api/internal/artifacts/configurations/DefaultConfiguration.java
Comment (jvandort) [reply platforms/software/dependency-management/src/main/java/org/gradle/api/internal/artifacts/configurations/DefaultConfiguration.java:915]: It made testing a little easier in `TasksFromProjectDependenciesTest` to avoid needing mocks

File: platforms/software/dependency-management/src/main/java/org/gradle/api/internal/artifacts/configurations/TasksFromProjectDependencies.java
Comment (jvandort) [reply platforms/software/dependency-management/src/main/java/org/gradle/api/internal/artifacts/configurations/TasksFromProjectDependencies.java:54]: Made this private & static. Removed the field. Deleted the test since it really didn't test anything

File: subprojects/core/src/main/java/org/gradle/api/internal/artifacts/CachingDependencyResolveContext.java
Comment (jvandort) [reply subprojects/core/src/main/java/org/gradle/api/internal/artifacts/CachingDependencyResolveContext.java:63]: Yes likely but I wanted to avoid messing with this. This class is deprecated and will be removed

File: subprojects/core/src/main/java/org/gradle/api/internal/artifacts/dependencies/DefaultFileCollectionDependency.java
Comment (jvandort) [reply subprojects/core/src/main/java/org/gradle/api/internal/artifacts/dependencies/DefaultFileCollectionDependency.java:104]: Yes

File: subprojects/core/src/main/java/org/gradle/api/internal/artifacts/dependencies/DefaultFileCollectionDependency.java
Comment (tresat) [reply subprojects/core/src/main/java/org/gradle/api/internal/artifacts/dependencies/DefaultFileCollectionDependency.java:91]: For sure the stacktrace.

Would the targetComponentId help at all?  That seems like it would always be present except in tests, right?

Comment (jvandort) [reply subprojects/core/src/main/java/org/gradle/api/internal/artifacts/dependencies/DefaultFileCollectionDependency.java:91]: I think it is often not present. That first constructor gets called by the instantiator, so its uses do not appear in the IDE

Comment (jvandort) [reply subprojects/core/src/main/java/org/gradle/api/internal/artifacts/dependencies/DefaultFileCollectionDependency.java:91]: We can print the target component ID but in practice I think the stack trace will guide the user to a solution in almost all cases.

File: platforms/core-configuration/configuration-cache/src/integTest/groovy/org/gradle/configurationcache/isolated/IsolatedProjectsAccessFromGroovyDslIntegrationTest.groovy
Comment (big-guy) [inline platforms/core-configuration/configuration-cache/src/integTest/groovy/org/gradle/configurationcache/isolated/IsolatedProjectsAccessFromGroovyDslIntegrationTest.groovy:693]: What's accessing the build dependencies here?

Comment (jvandort) [reply platforms/core-configuration/configuration-cache/src/integTest/groovy/org/gradle/configurationcache/isolated/IsolatedProjectsAccessFromGroovyDslIntegrationTest.groovy:693]: Thats the purpose of this test. It calls `<whatever>.buildDependencies.getDependencies(null)` for many <whatevers>, with a project dependency being one of them

File: platforms/documentation/docs/src/docs/userguide/releases/upgrading/upgrading_version_8.adoc
Comment (jvandort) [reply platforms/documentation/docs/src/docs/userguide/releases/upgrading/upgrading_version_8.adoc:58]: This is not a method users generally call directly, but via a `dependsOn` on their task. I've added a `dependsOn` and a comment stating that

File: platforms/documentation/docs/src/docs/userguide/releases/upgrading/upgrading_version_8.adoc
Comment (jvandort) [reply platforms/documentation/docs/src/docs/userguide/releases/upgrading/upgrading_version_8.adoc:90]: Added asserts in the first (deprecated) example to show these methods resolve to files and split this up into two tasks

File: platforms/software/dependency-management/src/main/java/org/gradle/api/internal/artifacts/dependencies/DefaultProjectDependency.java
Comment (jvandort) [reply platforms/software/dependency-management/src/main/java/org/gradle/api/internal/artifacts/dependencies/DefaultProjectDependency.java:114]: Added the project name to the message

File: platforms/software/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/SelfResolvingDependencyIntegrationTest.groovy
Comment (jvandort) [reply platforms/software/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/SelfResolvingDependencyIntegrationTest.groovy:52]: Would rather leave this as-is to reduce duplication. We'll be deleting the deprecated parts of this test soon enough

**Review by tresat — APPROVED** [review]

**Review by big-guy — APPROVED** [review]

**Comment by jvandort** [comment]
@bot-gradle test bd

**Comment by jvandort** [comment]
@bot-gradle test and merge

**Pull request merged**
